### PR TITLE
Integrate RTM with I18n-tasks

### PR DESCRIPTION
**Pull request by baisa**
Add i18n-tasks to Rails Translation Manager 

Add wrappers around `add-missing` & `normalize` tasks to be able to use them and have a common interface for both Gems.

[Trello card](https://trello.com/c/6RSfGJJa/2685-add-i18n-tasks-to-rails-translation-manager-5)

### CODE REVIEW

File: rails_translation_manager.gemspec [hunk 1]
```diff
@@ -21,7 +21,7 @@ Gem::Specification.new do |spec|
   spec.add_dependency "rails-i18n"
   spec.add_dependency "activesupport"
 
-  spec.add_development_dependency "bundler", "~> 1.7"
+  spec.add_development_dependency "bundler"
```
Comment (ChrisBAshton) [inline rails_translation_manager.gemspec:24]: Could you explain in the commit message why this has been changed? (It looks harmless enough to me though!)

**Review by ChrisBAshton — COMMENTED** [review]
This is looking good so far, @baisa 👍  I think a requirement may have been missed though. From the card:

> Something to be aware of is that i18n-tasks `add-missing` generates whitespace for the keys that are added. Ideally we don't want this to happen and have removed the whitespace with `find <path-to-locales> -type f -print0 | xargs -0 perl -p`.

So I think you'll need to modify the `normalize` task to ensure it removes the extra whitespace - and there should be a corresponding RSpec test for that.

File: lib/tasks/translation.rake [hunk 2]
```diff
@@ -83,4 +83,19 @@ namespace :translation do
     end
   end
 
+  desc "Add missing translations"
+  task(:add_missing, [:locale] => [:environment]) do |t, args|
+    locale = args[:locale] || "en"
+    require 'i18n/tasks/cli'
+
+    I18n::Tasks::CLI.start(["add-missing", "-l", locale])
```
Comment (1pretz1) [inline lib/tasks/translation.rake:91]: We probably want to make the `-l` flag optional, incase we want to add keys to multiple locales. There is also another flag `--nil-value` we will want to use, otherwise I18n-tasks tries to fill in values for all the keys it adds. Sorry, I should've mentioned the exact flags we intent to use in the Trello card.

File: lib/tasks/translation.rake [hunk 3]
```diff
@@ -83,4 +83,19 @@ namespace :translation do
...
+  end
+
+  desc "Normalize translations"
+  task(:normalize, [:locale] => [:environment]) do |t, args|
+    locale = args[:locale] || "en"
+    require 'i18n/tasks/cli'
+
+    I18n::Tasks::CLI.start(["normalize", "-l", locale])
```
Comment (1pretz1) [inline lib/tasks/translation.rake:99]: Same comment as above for making the `-l` flag optional.

**Comment by baisa** [comment]
> This is looking good so far, @baisa 👍 I think a requirement may have been missed though. From the card:
> 
> > Something to be aware of is that i18n-tasks `add-missing` generates whitespace for the keys that are added. Ideally we don't want this to happen and have removed the whitespace with `find <path-to-locales> -type f -print0 | xargs -0 perl -p`.
> 
> So I think you'll need to modify the `normalize` task to ensure it removes the extra whitespace - and there should be a corresponding RSpec test for that.

@ChrisBAshton this change is now implemented.

File: spec/rails_translation_manager/cleaner_spec.rb [hunk 4]
```diff
@@ -0,0 +1,17 @@
+require 'spec_helper'
+
+RSpec.describe RailsTranslationManager::Cleaner do
+
+  before(:each) do
+    FileUtils.cp('spec/locales/cleaner/with_whitespace.yml', 'spec/locales/cleaner/tmp')
```
Comment (ChrisBAshton) [inline spec/rails_translation_manager/cleaner_spec.rb:6]: By convention, we usually make use of the `tmp` directory at the root of the repository. I'd be tempted to make a `tmp/.gitkeep` and [add the folder to `.gitignore`](https://github.com/alphagov/content-publisher/blob/61181bca488aea8855b56d3582bef55b426d5ff7/.gitignore#L12). Then you don't have to worry about calling `rm_rf` afterwards.

I'd also rename this `tmp` => `tmp.yml`. It doesn't matter a lot, but the missing file extension threw me for a moment, as it's not clear if you're working with a file or a directory.

File: spec/rails_translation_manager/cleaner_spec.rb [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+  end
+
+  after(:each) do
+    FileUtils.rm_rf("spec/locales/cleaner/tmp", secure: true)
+  end
+
+  it 'removes whitespace at the end of the line' do
+    described_class.new('spec/locales/cleaner/tmp').clean
+    expect(FileUtils.compare_file('spec/locales/cleaner/tmp', 'spec/locales/cleaner/clean.yml')).to be_truthy
```
Comment (ChrisBAshton) [inline spec/rails_translation_manager/cleaner_spec.rb:15]: Nice test 👍

File: lib/tasks/translation.rake [hunk 6]
```diff
@@ -89,6 +89,7 @@ namespace :translation do
     require 'i18n/tasks/cli'
 
     locale ? I18n::Tasks::CLI.start(["add-missing", "--nil-value", "-l", locale]) : I18n::Tasks::CLI.start(["add-missing", "--nil-value"])
+    RailsTranslationManager::Cleaner.new(Rails.root.join("config", "locales")).clean
```
Comment (ChrisBAshton) [inline lib/tasks/translation.rake:92]: You've tested the `RailsTranslationManager::Cleaner` behaviour, which is great 👍 I think ideally we'd also have a test that calling these rake tasks triggers a call to `RailsTranslationManager::Cleaner`. But I don't think there are any existing tests for the Rake tasks, so this feels a bit out of scope.

File: rails_translation_manager.gemspec [hunk 7]
```diff
@@ -25,4 +25,5 @@ Gem::Specification.new do |spec|
   spec.add_development_dependency "rake", "~> 10.0"
   spec.add_development_dependency "minitest"
   spec.add_development_dependency 'rspec'
+  spec.add_development_dependency 'pry'
```
Comment (ChrisBAshton) [inline rails_translation_manager.gemspec:28]: Could this be moved into a separate commit, as it's unrelated to this change? But I agree it's worth adding - though you may want to consider switching to [byebug](https://rubygems.org/gems/byebug) instead, which is the one we seem to be switching to.

File: lib/tasks/translation.rake [hunk 8]
```diff
@@ -85,17 +85,17 @@ namespace :translation do
 
   desc "Add missing translations"
   task(:add_missing, [:locale] => [:environment]) do |t, args|
-    locale = args[:locale] || "en"
+    locale = args[:locale]
     require 'i18n/tasks/cli'
 
-    I18n::Tasks::CLI.start(["add-missing", "-l", locale])
+    locale ? I18n::Tasks::CLI.start(["add-missing", "--nil-value", "-l", locale]) : I18n::Tasks::CLI.start(["add-missing", "--nil-value"])
```
Comment (ChrisBAshton) [inline lib/tasks/translation.rake:91]: This could be [DRY](https://riptutorial.com/ruby-on-rails/example/3932/don-t-repeat-yourself--dry-)'d up by constructing the arguments array in advance:

```ruby
i18n_args = ["add-missing", "--nil-value"]
i18n_args << ["-l", locale] if locale

I18n::Tasks::CLI.start(i18n_args)
```

Could you explain in the commit message why we need the `--nil-value` flag?

**Review by ChrisBAshton — COMMENTED** [review]
Thanks, this is looking good, @baisa! A few comments below - and I also suggest changing the base of the PR to `add-locale-checker` (instead of `master`) as we only want to review your commits in isolation, not Peter's commits. When this PR is approved ✅, we can look at how best to get it merged. (Basically want to avoid merging Peter's changes until they've been approved too!)

File: spec/rails_translation_manager/cleaner_spec.rb (hunk 4 above)
Comment (baisa) [reply spec/rails_translation_manager/cleaner_spec.rb]: @ChrisBAshton I have created tmp/.gitkeep and removed after(:each) block.

File: lib/tasks/translation.rake (hunk 6 above)
Comment (baisa) [reply lib/tasks/translation.rake:95]: @ChrisBAshton should we create a new card to write those specs?

File: rails_translation_manager.gemspec (hunk 7 above)
Comment (baisa) [reply rails_translation_manager.gemspec]: @ChrisBAshton Have added byebug in separate commit.

File: lib/tasks/translation.rake (hunk 8 above)
Comment (baisa) [reply lib/tasks/translation.rake]: @ChrisBAshton I have applied it to add_missing task, and left normalize task unchanged (it doesn’t have  "--nil-value" flag so is more readable than add_missing).

File: spec/rails_translation_manager/cleaner_spec.rb (hunk 4 above)
Comment (ChrisBAshton) [reply spec/rails_translation_manager/cleaner_spec.rb]: Thanks 👍 Ah I think I'd misunderstood the code before. I assumed you were copying the `with_whitespace.yml` file to a new file called `tmp`, but I think you were actually copying the file into a _directory_ called `tmp`. And then the `RailsTranslationManager::Cleaner` class takes a _directory_ (not a file) as an argument. I understand now!

File: lib/tasks/translation.rake [hunk 9]
```diff
@@ -88,7 +88,11 @@ namespace :translation do
     locale = args[:locale]
     require 'i18n/tasks/cli'
 
-    locale ? I18n::Tasks::CLI.start(["add-missing", "--nil-value", "-l", locale]) : I18n::Tasks::CLI.start(["add-missing", "--nil-value"])
+    i18n_args = ["add-missing", "--nil-value"]
+    i18n_args << ["-l", locale] if locale
```
Comment (ChrisBAshton) [inline lib/tasks/translation.rake:92]: This change has ended up in the wrong commit - could you double-check? Thanks!

File: lib/tasks/translation.rake (hunk 6 above)
Comment (ChrisBAshton) [reply lib/tasks/translation.rake:95]: We wouldn't normally, as it would never get prioritised. It's more something I would hope someone would come in and add at the same time as adding a feature.

What I was trying to say (badly!) was that it would be great if you could add some tests for the rake tasks themselves, because there's a small gap in coverage here - but you don't need to, as there's not yet a precedent in this repo for testing rake tasks. So if you fancy the challenge, go for it, otherwise feel free to ignore my comment 😄

Comment (baisa) [reply lib/tasks/translation.rake:95]: @ChrisBAshton Makes sense, I will go ahead and add them!

File: lib/tasks/translation.rake [hunk 10]
```diff
@@ -85,17 +85,19 @@ namespace :translation do
...
   end
 
   desc "Normalize translations"
   task(:normalize, [:locale] => [:environment]) do |t, args|
-    locale = args[:locale] || "en"
+    locale = args[:locale]
     require 'i18n/tasks/cli'
 
-    I18n::Tasks::CLI.start(["normalize", "-l", locale])
+    locale ? I18n::Tasks::CLI.start(["normalize", "-l", locale]) : I18n::Tasks::CLI.start(["normalize"])
```
Comment (ChrisBAshton) [inline lib/tasks/translation.rake:102]: For consistency with the other rake task, I think I'd follow the same `i18n_args << ["-l", locale] if locale` approach here too.

OR, if you prefer, you could maybe make a helper method so that we can just call like this?

```ruby
I18n::Tasks::CLI.start(with_optional_locale(["normalize"]))
```

File: rails_translation_manager.gemspec [hunk 11]
```diff
@@ -24,5 +24,5 @@ Gem::Specification.new do |spec|
   spec.add_development_dependency "bundler"
   spec.add_development_dependency "rake", "~> 10.0"
   spec.add_development_dependency "minitest"
-  spec.add_development_dependency "rspec"
+  spec.add_development_dependency 'rspec'
```
Comment (ChrisBAshton) [inline rails_translation_manager.gemspec]: I think this came out of a rebase that went wrong, as Peter's PR originally used single quotes and then changed to double quotes for consistency with the rest of the file. I see you switch back to double quotes in the next commit. Could you fix this up here too so that there's no change overall?

File: lib/tasks/translation.rake [hunk 12]
```diff
@@ -85,17 +85,19 @@ namespace :translation do
 
   desc "Add missing translations"
   task(:add_missing, [:locale] => [:environment]) do |t, args|
-    locale = args[:locale] || "en"
+    locale = args[:locale]
     require 'i18n/tasks/cli'
 
-    I18n::Tasks::CLI.start(["add-missing", "-l", locale])
+    i18n_args = ["add-missing", "--nil-value"]
```
Comment (ChrisBAshton) [inline lib/tasks/translation.rake:91]: You've documented why we need the `--nil-value` flag in 4fba60d0426645a084add510f834314e42377792, but the change is introduced here (48a63d9675bfd970f185b1f306c9109d70fab192). Could you move the explanation into this commit?

File: lib/tasks/translation.rake [hunk 13]
```diff
@@ -1,4 +1,5 @@
 require "rails_translation_manager"
+require 'i18n/tasks/cli'
```
Comment (ChrisBAshton) [inline lib/tasks/translation.rake:2]: This change could be fixed up to an earlier commit (i.e. the commit where these rake tasks were first introduced). And also switch to double quotes for consistency with the line above.

File: lib/tasks/translation.rake [hunk 14]
```diff
@@ -83,4 +83,19 @@ namespace :translation do
     end
   end
 
+  desc "Add missing translations"
+  task(:add_missing, [:locale] => [:environment]) do |t, args|
+    locale = args[:locale] || "en"
+    require 'i18n/tasks/cli'
```
Comment (ChrisBAshton) [inline lib/tasks/translation.rake]: I see in 78fb156499251dcc8e41bb64ee50764d2e254f1e "Add tests for translations tasks" we first add the i18n-tasks gem. Should that change be moved to here? I imagine without it, these rake tasks would fail.

File: spec/tasks/translation_spec.rb [hunk 15]
```diff
@@ -0,0 +1,38 @@
...
+  before do
+    fake_rails = double()
+    fake_rails.stub(:root) { Pathname.new('spec') }
+    stub_const("Rails", fake_rails)
+  end
+
+  describe 'translation:add_missing', type: :task do
+    let(:task) { Rake::Task["translation:add_missing"] }
+  
+    it 'add missing task is executed' do
```
Comment (ChrisBAshton) [inline spec/tasks/translation_spec.rb:14]: This should read like a sentence, i.e. "it is executed":

```suggestion
    it 'is executed' do
```

File: spec/tasks/translation_spec.rb [hunk 16]
```diff
@@ -0,0 +1,38 @@
...
+      allow(I18n::Tasks::CLI).to receive(:start)
+      task.execute
+      expect(I18n::Tasks::CLI).to have_received(:start)
+    end
+  end
+
+  describe 'translation:normalize', type: :task do
+    let(:task) { Rake::Task["translation:normalize"] }
+  
+    it 'normalize task is executed' do
```
Comment (ChrisBAshton) [inline spec/tasks/translation_spec.rb:28]:
```suggestion
    it 'is executed' do
```

**Review by ChrisBAshton — COMMENTED** [review]
This is looking great, @baisa! 👏 Thanks for addressing my earlier comments - and thanks for adding the rake task tests, which goes above and beyond what's strictly needed for the PR ⭐ 

Just a couple of really minor comments left, mostly around which code changes appear in which commits. Do reach out to me if I need to explain any of these in person!

File: lib/tasks/translation.rake (hunk 12 above)
Comment (baisa) [reply lib/tasks/translation.rake]: I have moved the commit description.

File: lib/tasks/translation.rake (hunk 13 above)
Comment (baisa) [reply lib/tasks/translation.rake]: @ChrisBAshton have moved it to the first commit

File: lib/tasks/translation.rake [hunk 17]
```diff
@@ -83,4 +84,19 @@ namespace :translation do
...
+
+    I18n::Tasks::CLI.start(["add-missing", "-l", locale])
+  end
+
+  desc "Normalize translations"
+  task(:normalize, [:locale] => [:environment]) do |t, args|
+    locale = args[:locale] || "en"
+    require 'i18n/tasks/cli'
```
Comment (ChrisBAshton) [inline lib/tasks/translation.rake]: These `require` calls can be removed now that you `require` on line 2.

File: lib/tasks/translation.rake [hunk 18]
```diff
@@ -87,21 +88,16 @@ namespace :translation do
   desc "Add missing translations"
   task(:add_missing, [:locale] => [:environment]) do |t, args|
     locale = args[:locale]
-    require 'i18n/tasks/cli'
 
-    i18n_args = ["add-missing", "--nil-value"]
-    i18n_args << ["-l", locale] if locale
-
-    I18n::Tasks::CLI.start(i18n_args.flatten)
+    I18n::Tasks::CLI.start(TranslationHelper.new(["add-missing", "--nil-value"], locale).with_optional_locale)
```
Comment (ChrisBAshton) [inline lib/tasks/translation.rake:92]: You can now pass `args[:locale]` directly, and remove the `locale=` variable on line 90, as it only gets used here. Same comment for the other rake task.

File: spec/tasks/translation_spec.rb [hunk 19]
```diff
@@ -0,0 +1,38 @@
...
+      task.execute
+      expect(I18n::Tasks::CLI).to have_received(:start)
+    end
+  end
+
+  describe 'translation:normalize', type: :task do
+    let(:task) { Rake::Task["translation:normalize"] }
+  
+    it 'is executed' do
+      expect { task.execute }.to_not output.to_stdout
```
Comment (ChrisBAshton) [inline spec/tasks/translation_spec.rb:29]: This surprised me - do we not expect to see any output from this command?
Maybe it's worth adding a `puts "Done!"` at the end of the Rake task, so that we can check that the output contains `Done!`?

File: lib/tasks/translation_helper.rb [hunk 20]
```diff
@@ -0,0 +1,11 @@
+class TranslationHelper
```
Comment (ChrisBAshton) [inline lib/tasks/translation_helper.rb:1]: Nice use of a helper 👍

File: lib/tasks/translation.rake [hunk 21]
```diff
@@ -87,21 +88,16 @@ namespace :translation do
...
     RailsTranslationManager::Cleaner.new(Rails.root.join("config", "locales")).clean
   end
 
   desc "Normalize translations"
   task(:normalize, [:locale] => [:environment]) do |t, args|
     locale = args[:locale]
-    require 'i18n/tasks/cli'
 
-    locale ? I18n::Tasks::CLI.start(["normalize", "-l", locale]) : I18n::Tasks::CLI.start(["normalize"])
+    I18n::Tasks::CLI.start(TranslationHelper.new(["normalize"], locale).with_optional_locale)
```
Comment (ChrisBAshton) [inline lib/tasks/translation.rake:100]: Does this work? It looks like `I18n::Tasks::CLI.start` takes an array, but we're passing it a `TranslationHelper` object.

Don't we need to define a `to_a` on `TranslationHelper` to return the `@task_options` array?

File: spec/tasks/translation_spec.rb [hunk 22]
```diff
@@ -0,0 +1,38 @@
...
+    let(:task) { Rake::Task["translation:add_missing"] }
+  
+    it 'is executed' do
+      expect { task.execute }.to output.to_stdout
+    end
+
+    it 'triggers i18n task' do
+      allow(I18n::Tasks::CLI).to receive(:start)
+      task.execute
+      expect(I18n::Tasks::CLI).to have_received(:start)
```
Comment (ChrisBAshton) [inline spec/tasks/translation_spec.rb:21]: We should also check that it's receiving the right argument, e.g.

```suggestion
      expect(I18n::Tasks::CLI).to have_received(:start).with(["add-missing"])
```

**Review by ChrisBAshton — COMMENTED** [review]
Thanks for fixing up the commits! 🎉 

Just a couple of questions below - all minor except the `TranslationHelper` testing, which I suspect might need a second look?

File: spec/tasks/translation_spec.rb (hunk 19 above)
Comment (baisa) [reply spec/tasks/translation_spec.rb:29]: @ChrisBAshton For `normalize` task this is expected behaviour - task doesn’t return any output, in opposite to `add_missing`. I have been additionally testing this by running tasks on `collections` app.

File: lib/tasks/translation.rake (hunk 21 above)
Comment (baisa) [reply lib/tasks/translation.rake:100]: @ChrisBAshton Here, on `TranslationHelper` instance we call `with_optional_locale` method which returns an array.

File: lib/tasks/translation.rake (hunk 18 above)
Comment (baisa) [reply lib/tasks/translation.rake]: @ChrisBAshton good point, it is fixed!

File: spec/tasks/translation_spec.rb (hunk 22 above)
Comment (baisa) [reply spec/tasks/translation_spec.rb:21]: @ChrisBAshton Have added those, thank you 👍

Comment (ChrisBAshton) [reply spec/tasks/translation_spec.rb:21]: Thanks! You can remove line 21 now, we only want the more specific line 22.

File: spec/tasks/translation_spec.rb (hunk 19 above)
Comment (ChrisBAshton) [reply spec/tasks/translation_spec.rb:30]: Fair enough 👍

**Review by ChrisBAshton — COMMENTED** [review]
Thanks for the continued hard work on this, @baisa. I think this is ready to go now!

Peter has made some changes to his branch, so you're now seeing conflicts on yours. We should have a Slack chat between the three of us to decide which PR to merge first.

File: lib/tasks/translation.rake (hunk 21 above)
Comment (ChrisBAshton) [reply lib/tasks/translation.rake]: Thanks for explaining - I forgot about the implicit return!
